rfc/languages-in-w3s.md

```diff
@@ -0,0 +1,84 @@
+- Modern javascript is a highly capable language, if you unify a your team on a single coding style, linter, tool chain, package manager, approach to typing, and transpilation system as needed. Unfortunately, almost every one of the above has multiple options and the ecosystem is an incoherent mess.
+- Javascript's single threaded asynchronous programming model turns out to actually be pretty decent for writing web servers too.
+- A lot of devs are used to deploying code on centralized services with built in javascript/node support and forget that distributing binaries is still the best way to get code to machines in a truly distributed system like the one we're building. Javascript's binary story kind of sucks. And, while JS runtime(s) are pretty optimized, JS is still pretty unpredictable in terms of memory usage, and to a lesser extent performance. Javascript, at least in its node.js/browser form, also has a pretty big size runtime(s).
+- Every hiring manager loves javascript cause it's easy to hire javascript developers. The main problem with this approach is there are so many kinds of javascript development and ways to write javascript that hiring a javascript developer is not a gaurantee they'll be able to write javascript the way your team uses it.
+
+#### Some Notes On Our Javascript
+
+- Thankfully this team has unified on a coding style, linter, tool chain, package manager, and approach to typing. +1 on that.
+- We have not however documented why we made the choices we did. We should do so ASAP. In particular types in JSDoc style comments and why we do it needs documentation. This includes our sporadic dropping into Typescript and why we do it. (For example: 'why is there an api.js and an api.ts and the js file just says `export {}` -- the answer is 'cause modules' but is EXTREMELY not clear)
+- The w3up repo is huge, and honestly, even if you know javascript, there's a lot to get up to speed on. We should have a guide to the code.
+
+### Go
+
+- Go is a highly performant garbage collected language like Java, but is much simpler than Java
+- Go has an extremely powerful concurrency model with green threads
+- These two components together make Go one of the most powerful languages for writing servers of any kind. The challenge of server programming is not super fast compute, but super high concurrency execution. Go excels at this, which matters cause our project has a lot of servers involved.
+- Go's ability to do binary distributions by default is a major plus for kind of project.
+- By and large, people get stuff done quickly in Go because Go is extremely simple. It's very fast to learn, easy to read, and it all looks the same. The only problem is its extremely boring. (Note from Hannah: When I first started writing Go I hated it. I thought someone took the joy out of programming. Now I love it cause I get my joy from what I'm building, not my programming language)
```

Comment:
> only problem is its extremely boring

I have found Go to be one of the most interesting languages from my first use, and up to the present. It is interesting because it lets you do everything you want efficiently and elegantly, and _still_ stays out of your way. It has a nearly perfect balance, offering what is most needed and useful but nothing more to complicate things. I have not yet encountered another language that gets the balance so close to just right for most programming projects, particularly, those involving network and systems programming.

You do not need to choose a package manager, transpilation tools, or even a coding style - those are not interesting problems to solve. There is no requirement for your target execution environment to have the correct version of a particular runtime and dependencies. You can focus on solving concurrency problems without having to solve the system-dependent problem of parallelism (managing OS threads running across multiple cores). Go is interesting (and fun) because it avoids the annoying ceremony of how to go about doing things, and lets you just do things, right away.

... but, as previously stated, the engineering problem to solve is primarily what must determine which tools are best suited to solve it.